app/views/layouts/menus/_student_sub_menu.html.erb

```diff
@@ -53,13 +53,28 @@
       <% end %>
 
       <% @submission = @grouping.nil? ? nil : @grouping.current_submission_used %>
-      <% @result = @submission.nil? ? nil : @submission.get_original_result %>
+      <% @result = @submission.nil? ? nil : @submission.get_visible_result %>
       <% if !@grouping.nil? && !@submission.nil? && !@result.nil? %>
-        <li class='<%= "active" if controller.controller_name == 'results' &&
-          !@result.is_a_review? %>'>
-          <%= link_to Result.model_name.human.pluralize,
-                      view_marks_course_result_path(@current_course, @result) %>
-        </li>
+        <% active = controller.controller_name == 'results' && !@result.is_a_review? %>
+        <% if allowed_to?(:view?, @result, context: {view_token: session["view_token"]&.[](@result&.id&.to_s)}) %>
+          <!-- check if a token not required or if the current user has entered the token already this session -->
+          <li class='<%= "active" if active %>'>
+            <%= link_to Result.model_name.human.pluralize,
+                        view_marks_course_result_path(@current_course, @result) %>
+          </li>
+        <% elsif allowed_to? :view_token_check?, @result %>
+          <li class='<%= "active" if active %>'>
+            <%= link_to Result.model_name.human.pluralize, '#', id: "submit-view-token" %>
+          </li>
+          <div id="submit-view-token-modal"></div>
```

Comment:
Putting the `div` here breaks the horizontal flow of the `li`s across the page (specifically the Peer Reviews tab below gets shifted to a new line). You can move the div elsewhere but keep the javascript here (or move both).